Feature/be s3 upload service

vic2019 opened the pull request.

Commit:
Add api to upload file to s3

Commit:
Edit typo in .env.template

Commit:
Uncomment middleware in routes/files.js

maismin commented on server/app/routes/files.js:
```diff
@@ -0,0 +1,12 @@
+
+const router = express.Router();
+
+router.use(expressFileUpload());
```
The expressFileUpload middleware needs to go inside app.js and BEFORE the endpoint routes, otherwise the routers won't have access to it.

maismin commented on server/app/routes/files.js:
```diff
@@ -0,0 +1,12 @@
+const router = express.Router();
+
+router.use(expressFileUpload());
+router.post('/upload', protect, upload);
```
Instead of '/upload', perhaps '/' is good enough?
With '/upload' the API looks like this:
``
POST /api/v1/files/upload
``
This says that we're creating an upload, which doesn't sound quite right

I think
``
POST /api/v1/files
``
makes more sense, as it says we're creating a file

maismin reviewed (changes requested):
The code looks very clean. Just remember to move the express middleware to the app.js and fix the error message. Try to leave some comments so that others seeing your code for the first time have an idea of what's going on, and for yourself when you come back to look at the code later.

maismin commented on server/app/controllers/files.js:
```diff
@@ -0,0 +1,34 @@
+const { BUCKET_NAME, UPLOAD_DIRNAME } = process.env;
+
+const upload = asyncHandler(async (req, res, next) => {
+  if (!req.files || !req.files.file) throw Error('File is undefined');
```
Instead of throwing an error, I created an ErrorResponse class that you can pass to the next() method. 
``
return next(new ErrorResponse('File is undefined', 400))
``
You can pass an error message and the status code.

Commit:
- Move express-fileupload to app.js
- Change route to POST /api/v1/files
- Change `throw Error` to `next(new ErrorResponse(... ))`

maismin commented on server/app/controllers/files.js:
```diff
@@ -0,0 +1,38 @@
+
+const upload = asyncHandler(async (req, res, next) => {
+  if (!req.files || !req.files.file) {
+    next(new ErrorResponse('File is undefined', 400));
+    return;
```
Having the ``return`` below the ``next()`` function means that the upload function will pass the ErrorResponse to the next() function and then return back to the upload function, which then ends. There is too much jumping around in terms of the function calls. So put the ``return`` next to the ``next()`` call

``
return next(new ErrorResponse('File is undefined', 400));
``

EDIT: After doing some some research and thinking, I don't think there's a difference in whether or not the return is next to the `next()` function except for styles. Having the `return` on the next line indicates that we're not returning anything other than ending the current execution context, so it makes more sense here.

vic2019 replied:
I agree that having return on the same line makes more sense. The only reason I put it on the next line is because I would get a linter warning about arrow function not returning anything. That warning doesn't always come up for every `return next(new ErrorResponse( ...`, so I'm not sure why. But we can change it later.

vic2019 replied:
Turned out that I was missing a return statement at the end of the block, which was causing the problem. I've added the change as you commented. Thanks!
